be/src/storage/compaction_manager.cpp

```diff
@@ -575,8 +574,47 @@ std::unordered_set<CompactionTask*> CompactionManager::get_running_task(const Ta
     return res;
 }
 
+int32_t CompactionManager::compute_max_compaction_task_num() const {
+    int32_t max_task_num = 0;
+    // new compaction framework
+    if (config::base_compaction_num_threads_per_disk >= 0 && config::cumulative_compaction_num_threads_per_disk >= 0) {
+        max_task_num = static_cast<int32_t>(
+                StorageEngine::instance()->get_store_num() *
+                (config::cumulative_compaction_num_threads_per_disk + config::base_compaction_num_threads_per_disk));
+    } else {
+        // When cumulative_compaction_num_threads_per_disk or config::base_compaction_num_threads_per_disk is less than 0,
+        // there is no limit to _max_task_num if max_compaction_concurrency is also less than 0, and here we set maximum value to be 20.
+        max_task_num = std::min(20, static_cast<int32_t>(StorageEngine::instance()->get_store_num() * 5));
+    }
+
+    {
+        std::lock_guard lg(_compact_threads_mutex);
+        if (_max_compaction_concurrency > 0 && _max_compaction_concurrency < max_task_num) {
+            max_task_num = _max_compaction_concurrency;
+        }
+    }
+
+    return max_task_num;
+}
+
+void CompactionManager::set_max_compaction_concurrency(int threads_num) {
```

Comment:
this function seems useless, since both `_max_compaction_concurrency` and `_max_task_num` come from `max_threads`

Reply:
1 max_threads is come from config::max_compaction_concurrency，_max_compaction_concurrency record the config, hherefore _max_compaction_concurrency must be set in the current context.
```
set_max_compaction_concurrency(max_threads)
```
2 _max_task_num is come form compute_max_compaction_concurrency

3 In original olap_server code, _max_task_num is initialized in main thread, there is no concurrency issue。Now, it need set_max_compaction_concurrency to avoid concurrency issue.